// tailwindcss/lib/util/dataTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    normalize: function() {
        return normalize;
    },
    url: function() {
        return url;
    },
    number: function() {
        return number;
    },
    percentage: function() {
        return percentage;
    },
    length: function() {
        return length;
    },
    lineWidth: function() {
        return lineWidth;
    },
    shadow: function() {
        return shadow;
    },
    color: function() {
        return color;
    },
    image: function() {
        return image;
    },
    gradient: function() {
        return gradient;
    },
    position: function() {
        return position;
    },
    familyName: function() {
        return familyName;
    },
    genericName: function() {
        return genericName;
    },
    absoluteSize: function() {
        return absoluteSize;
    },
    relativeSize: function() {
        return relativeSize;
    }
});
const _color = require("./color");
const _parseBoxShadowValue = require("./parseBoxShadowValue");
const _splitAtTopLevelOnly = require("./splitAtTopLevelOnly");
let cssFunctions = [
    "min",
    "max",
    "clamp",
    "calc"
];
// Ref: https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Types
function isCSSFunction(value) {
    return cssFunctions.some((fn)=>new RegExp(`^${fn}\\(.*\\)`).test(value));
}
// These properties accept a `<dashed-ident>` as one of the values. This means that you can use them
// as: `timeline-scope: --tl;`
//
// Without the `var(--tl)`, in these cases we don't want to normalize the value, and you should add
// the `var()` yourself.
//
// More info:
// - https://drafts.csswg.org/scroll-animations/#propdef-timeline-scope
// - https://developer.mozilla.org/en-US/docs/Web/CSS/timeline-scope#dashed-ident
//
const AUTO_VAR_INJECTION_EXCEPTIONS = new Set([
    // Concrete properties
    "scroll-timeline-name",
    "timeline-scope",
    "view-timeline-name",
    "font-palette",
    // Shorthand properties
    "scroll-timeline",
    "animation-timeline",
    "view-timeline"
]);
function normalize(value, context = null, isRoot = true) {
    let isVarException = context && AUTO_VAR_INJECTION_EXCEPTIONS.has(context.property);
    if (value.startsWith("--") && !isVarException) {
        return `var(${value})`;
    }
    // Keep raw strings if it starts with `url(`
    if (value.includes("url(")) {
        return value.split(/(url\(.*?\))/g).filter(Boolean).map((part)=>{
            if (/^url\(.*?\)$/.test(part)) {
                return part;
            }
            return normalize(part, context, false);
        }).join("");
    }
    // Convert `_` to ` `, except for escaped underscores `\_`
    value = value.replace(/([^\\])_+/g, (fullMatch, characterBefore)=>characterBefore + " ".repeat(fullMatch.length - 1)).replace(/^_/g, " ").replace(/\\_/g, "_");
    // Remove leftover whitespace
    if (isRoot) {
        value = value.trim();
    }
    value = normalizeMathOperatorSpacing(value);
    return value;
}
/**
 * Add spaces around operators inside math functions
 * like calc() that do not follow an operator, '(', or `,`.
 *
 * @param {string} value
 * @returns {string}
 */ function normalizeMathOperatorSpacing(value) {
    let preventFormattingInFunctions = [
        "theme"
    ];
    let preventFormattingKeywords = [
        "min-content",
        "max-content",
        "fit-content",
        // Env
        "safe-area-inset-top",
        "safe-area-inset-right",
        "safe-area-inset-bottom",
        "safe-area-inset-left",
        "titlebar-area-x",
        "titlebar-area-y",
        "titlebar-area-width",
        "titlebar-area-height",
        "keyboard-inset-top",
        "keyboard-inset-right",
        "keyboard-inset-bottom",
        "keyboard-inset-left",
        "keyboard-inset-width",
        "keyboard-inset-height",
        "radial-gradient",
        "linear-gradientтключена, приложение не будет обновляться.Просмотр веб-страницТакже удалить данные браузера?Изменить браузер по умолчанию на:Microsoft Edge заблокировал этот файл, так как этот тип файла опасенMicrosoft Edge заблокировал этот файл, так как он является опаснымРасширения, приложения и темы из неизвестных источников могут нанести вред устройству. Microsoft Edge рекомендует устанавливать их только из $1Microsoft Edge заблокировал этот файл, так как он содержит вредоносную программуMicrosoft Edge рекомендует сканировать этот файл, так как он может быть опасенУзнайте, почему Microsoft Edge блокирует некоторые загрузкиДобавление в Microsoft Edge…Файл $1 может быть опасным, поэтому Microsoft Edge заблокировал его.Браузер Microsoft Edge заблокировал файл $1 как небезопасный.{NUM_DOWNLOAD, plural,
          =1 {В случае закрытия браузера возможна потеря данных.} one {В случае закрытия браузера возможна потеря данных.} few {В случае закрытия браузера возможна потеря данных.} many {В случае закрытия браузера возможна потеря данных.} other {В случае закрытия браузера возможна потеря данных.}}Internet Explorer все еще загружается. Вы можете потерять загрузку, если закроете браузер.Необходимо добавить ключ API (Майкрософт) в приложении, на веб-сайте или на сервере.Расширение "$1" добавлено в Microsoft EdgeТакже удалить данные из Microsoft Edge ($1)Microsoft Edge обнаружил, что "$1" содержит вредоносную программуMicrosoft Edge обнаружил, что эти элементы содержат вредоносные программы:Это расширение содержит вредоносную программу и является небезопасным. Удалите его из Microsoft Edge, чтобы оно больше не просматривало и не изменяло ваши данные (включая личные сведения) на посещаемых сайтах.Это расширение нарушает политику веб-магазина Microsoft Edge и может быть небезопасным. Удалите его из Microsoft Edge, чтобы оно больше не просматривало и не изменяло ваши данные (включая личные сведения) на посещаемых сайтах.Публикация этого расширения была отменена разработчиком, и оно может быть небезопасным. Удалите его из Microsoft Edge, чтобы оно больше не просматривало и не изменяло ваши данные (включая личные сведения) на посещаемых сайтах.Microsoft Edge рекомендует просмотреть это расширениеРазрешить расширению отображать запросы доступа на панели инструментов Microsoft EdgeЕсли вы выберете этот параметр, журнал браузера все еще может быть записан. Edge не может запретить расширению сохранять журнал браузера даже в режиме InPrivate.Удалить из Microsoft EdgeВ EdgeЧтобы сделать Microsoft Edge более безопасным, мы отключили некоторые расширения, которые могли быть добавлены без вашего ведома.Чтобы сделать Microsoft Edge более безопасным, мы отключили следующее расширение, которое могло быть добавлено без вашего ведома.Настройки и прочее (ALT+F). Доступно обновление.Настройки и прочее (ALT+F). Некоторые вопросы требуют вашего внимания.&Открыть в Microsoft EdgeО программе &Microsoft Edge&Доступно обновление&Рекомендуется выполнить обновление&ОбновитьПерезапустите, чтобы обновить &Microsoft EdgeРанее вы входили от имени $1. Войдите снова с помощью этой учетной записи.Предыдущий вход был выполнен от имени $1. Чтобы ваше избранное, журнал и другие настройки не были объединены с этой учетной записью, создайте собственную учетную запись Microsoft Edge.Вы вошли в Microsoft Edge.Вы вошли в систему от имени $1. Теперь вы можете получать доступ к избранному, журналу и другим настройкам на всех устройствах, где выполнен вход в систему.Пользователь $1 работал в Microsoft Edge.Продолжить в новом профиле Microsoft Edge?Переключиться на существующий профиль Microsoft Edge?Профиль Microsoft Edge с этой учетной записью уже существуетПользователь $1 уже вошел в этот профиль Microsoft Edge. Чтобы отделить ваши данные о просмотре веб-страниц, Microsoft Edge может создать ваш собственный профиль.Пользователь $1 уже вошел в свой профиль Microsoft Edge. Будет создан новый профиль Microsoft Edge для $2Будет создан новый профиль Microsoft Edge для $1Настройка нового профиля Microsoft EdgeНастройка профиля EdgeMicrosoft Edge использует вашу камеру и микрофон.Microsoft Edge использует ваш микрофон.Microsoft Edge использует вашу камеру.{NUM_DEVICES, plural,
            =0 {Расширение Microsoft Edge обращалось к устройствам HID}
            =1 {Расширение Microsoft Edge обращается к 1 устройству HID}
            one {Расширение Micro
