// sucrase/dist/esm/cli.js
/* eslint-disable no-console */
import commander from "commander";
import {glob} from "glob";
import {exists, mkdir, readdir, readFile, stat, writeFile} from "mz/fs";
import {dirname, join, relative} from "path";

import { transform} from "./index";











export default function run() {
  commander
    .description(`Sucrase: super-fast Babel alternative.`)
    .usage("[options] <srcDir>")
    .option(
      "-d, --out-dir <out>",
      "Compile an input directory of modules into an output directory.",
    )
    .option(
      "-p, --project <dir>",
      "Compile a TypeScript project, will read from tsconfig.json in <dir>",
    )
    .option("--out-extension <extension>", "File extension to use for all output files.", "js")
    .option("--exclude-dirs <paths>", "Names of directories that should not be traversed.")
    .option("-q, --quiet", "Don't print the names of converted files.")
    .option("-t, --transforms <transforms>", "Comma-separated list of transforms to run.")
    .option("--disable-es-transforms", "Opt out of all ES syntax transforms.")
    .option("--jsx-runtime <string>", "Transformation mode for the JSX transform.")
    .option("--production", "Disable debugging information from JSX in output.")
    .option(
      "--jsx-import-source <string>",
      "Automatic JSX transform import path prefix, defaults to `React.Fragment`.",
    )
    .option(
      "--jsx-pragma <string>",
      "Classic JSX transform element creation function, defaults to `React.createElement`.",
    )
    .option(
      "--jsx-fragment-pragma <string>",
      "Classic JSX transform fragment component, defaults to `React.Fragment`.",
    )
    .option("--keep-unused-imports", "Disable automatic removal of type-only imports/exports.")
    .option("--preserve-dynamic-import", "Don't transpile dynamic import() to require.")
    .option(
      "--inject-create-require-for-import-require",
      "Use `createRequire` when transpiling TS `import = require` to ESM.",
    )
    .option(
      "--enable-legacy-typescript-module-interop",
      "Use default TypeScript ESM/CJS interop strategy.",
    )
    .option("--enable-legacy-babel5-module-interop", "Use Babel 5 ESM/CJS interop strategy.")
    .parse(process.argv);

  if (commander.project) {
    if (
      commander.outDir ||
      commander.transforms ||
      commander.args[0] ||
      commander.enableLegacyTypescriptModuleInterop
    ) {
      console.error(
        "If TypeScript project is specified, out directory, transforms, source " +
          "directory, and --enable-legacy-typescript-module-interop may not be specified.",
      );
      process.exit(1);
    }
  } else {
    if (!commander.outDir) {
      console.error("Out directory is required");
      process.exit(1);
    }

    if (!commander.transforms) {
      console.error("Transforms option is required.");
      process.exit(1);
    }

    if (!commander.args[0]) {
      console.error("Source directory is required.");
      process.exit(1);
    }
  }

  const options = {
    outDirPath: commander.outDir,
    srcDirPath: commander.args[0],
    project: commander.project,
    outExtension: commander.outExtension,
    excludeDirs: commander.excludeDirs ? commander.excludeDirs.split(",") : [],
    quiet: commander.quiet,
    sucraseOptions: {
      transforms: commander.transforms ? commander.transforms.split(",") : [],
      disableESTransforms: commander.disableEsTransforms,
      jsxRuntime: commander.jsxRuntime,
      production: commander.production,
      jsxImportSource: commander.jsxImportSource,
      jsxPragma: commander.jsxPragma || "React.createElement",
      jsxFragmentPragma: commander.jsxFragmentPragma || "React.Fragment",
      keepUnusedImports: commander.keepUnusedImports,
      preserveDynamicImport: commander.preserveDynamicImport,
      injectCreateRequireForImportRequire: commander.injectCreateRequireForImportRequire,
      enableLegacyTypeScriptModuleInterop: commander.enableLegacyTypescriptModuleInterop,
      enableLegacyBabel5ModuleInterop: commander.enableLegacyBabel5ModuleInterop,
    },
  }$1, жауап беру, $2$2$1 bookmarkАралық сақтағыштың кескінін іздеуАралық сақтағыштың мәтінін іздеу, $2Аралық сақтағыштың URL мекенжайын іздеу, $2$1 қазір ашық тұр. Tab пернесін, одан кейін Enter пернесін басып, ашық тұрған қойыншаға ауысыңызҚойыншаны ауыстыру түймесі, осы қойыншаға ауысу үшін Enter пернесін басыңыз$1, бірнеше әрекет қолжетімді, солардың арасында жылжу үшін, Tab пернесін басыңыз$1, іздеу үшін, Tab пернесін, содан кейін Enter пернесін басыңызІздеу: $1, type a query and press Enter to search$1, ұсынысты жою үшін Tab пернесін, содан кейін Enter пернесін басыңыз.«Ұсынысты жою» түймесі, жою үшін Enter пернесін басыңыз, $1Ұсыныстарды көрсетуҰсыныстарды жасыру"$1" бөлімін көрсету"$1" бөлімін жасыру"$1" бөлімі көрсетілген"$1" бөлімі жасырылған$1 = $2Блокталған мазмұнҚауіпті мазмұнКүдікті сайтТанымалҮздік сайттарыңызLinuxMacOSCrOSПланшетБасқа құрылғылар"$1" сұрауы бойынша жұмыс іздеу$1 ішінде "$2" іздеу"$1" сөзін теріп, $2 пернесін басыңызТек $1 ішінде іздеуWord құжатыPowerPoint құжатыExcel құжатыVisio құжатыFluid құжатыXML құжаты$2, $1, $3$2, $1, Бетбелгі, $3ЖасағанжұмысМекенжай жолағынан іздеу үшін осы таңбашаны таңдаңызҚажетті нәрсені жылдамырақ табыңызТікелей мекенжай жолағынан іздеу арқылы жұмысыңызды үнемдеңізБайқап көру үшін осы таңбашаны таңдаңыз"Жұмыс" сөзін теріп, Tab пернесін басыңыз$1, журналдағы соңғы іздеулер$1, $2 желісінде таңдаулы іздеу сұрауы$1, іздеуіңізге қатысты, $2 жүйесінен іздеу"$1" ұсынысын өшіру$2$1$2$1, жақында жабылған қойыншалардан$2$1, басқа құрылғылардағы қойыншаларданБайланыс қауіпсізОсы сайтқа байланысыңыз толығымен қауіпсіз емесБайланыс толықтай қауіпсіз емесОсы сайтқа қосылым қауіпсіз емесБайланыс қауіпсіз емесОсы сайтта зиянды бағдарлама барҚауіпті сайтБұл сайт күмәнді болып көрінедіОсы сайтта зиянды бағдарламалар барСіз кеңейтім бетін қарап жатырсызСіз веб-беттің көзін қарап жатырсызӘзірлеуші құралдары бетін көріп жатырсызMicrosoft Edge браузері қауіпсіз қосылу үшін бастапқы веб-бетті алды және оқылуын жеңілдету үшін оны жеңілдетті.Microsoft Edge браузері қауіпсіз емес қосылу үшін бастапқы веб-бетті алды және оқылуын жеңілдету үшін оны жеңілдетті.Иә, жалғастыру"$1" деп жазғыңыз келіп пе еді?Бұл сайт қауіпсіз емес сияқтыСіз жергілікті немесе бөлісілген файлды көрудесізСіздің ақпаратыңыз (мысалы, құпиясөздер немесе кредиттік карталарының нөмірлері) осы сайтқа жіберілген кезде жеке сақталады.Сіз әкімшімен берілген куәлік арқылы мазмұнға қатынастыңыз. Сіз $1 доменіне ұсынған деректерді әкімші ұстап алуы мүмкін.Шабуылшылар осы сайтта қарап отырған суреттерді көруі мүмкін және оларды өзгерту арқылы алдауы мүмкін.Бұл сайтта ешқандай құпия ақпарат енгізбеңіз (мысалы, құпиясөздер немесе кредиттік карталар). Ол шабуылшылармен ұрлануы мүмкін.Осы сайттағы шабуылшылар компьютерде қауіпті бағдарламаларды орнатуға тырысуы мүмкін. Олар ақпаратты (мысалы, фотосуреттер, құпиясөздер, хабарлар мен кредиттік карталарды) ұрлауы немесе жоюы мүмкін.Сіз кіруге әрекеттеніп жатқан сайттағы шабуылдаушылар құпиясөз, фотосуре
